src/LibHdb++MySQL.cpp

```diff
@@ -4217,18 +3049,105 @@ bool HdbPPMySQL::autodetect_column(string table_name, string column_name)
 	return false;
 }
 
+int HdbPPMySQL::cache_ID(const string &attr, const string &func_name)
+{
+	unordered_map<string,int>::iterator it = attr_ID_map.find(attr);
+	//if not already present in cache, look for ID in the DB
+	if(it == attr_ID_map.end())
+	{
+		int ID=-1;
+		string facility = get_only_tango_host(attr);
+		string attr_name = get_only_attr_name(attr);
+		find_attr_id(facility, attr_name, ID);
+		if(ID != -1)
+		{
+			attr_ID_map.insert(make_pair(attr,ID));
+			it = attr_ID_map.find(attr);
+		}
+		else
+		{
+			cout << __func__<< ": ID not found!" << endl;
+			Tango::Except::throw_exception(DATA_ERROR,"ID not found",func_name.c_str());
+		}
+	}
+	if(it == attr_ID_map.end())
+	{
+		cout << __func__<< ": ID not found for attr="<<attr << endl;
+		Tango::Except::throw_exception(DATA_ERROR,"ID not found",func_name.c_str());
+	}
+	return it->second;
+}
+
+bool HdbPPMySQL::cache_pstmt(const string &query, MYSQL_STMT **pstmt, uint32_t stmt_size/*TODO: handle max-prepared-stmt-count*/, const string &func_name)
+{
+	//pstmt = nullptr;
+	unsigned long mti = mysql_thread_id(dbp);
+	if(db_mti != mti)
+	{
+#ifdef _LIB_DEBUG
+		cout << func_name<< ": changed mysql_thread_id from " << db_mti << " to " << mti << endl;
+#endif
+		db_mti = mti;
+		unordered_map<string,MYSQL_STMT *>::iterator it_pstmt;
+		for(it_pstmt = pstmt_map.begin(); it_pstmt != pstmt_map.end(); it_pstmt++)
+		{
+			if (mysql_stmt_close(it_pstmt->second))
```

Comment:
ah ! Almost same code as line 733… what happened to the if(!cached) ?

Reply:
Here code is different: inside the caching function, closing all prepared statements when reconnection to server is detected. In other places it was closing only the prepared statement who failed in case it was not cached.